### DISCUSSION TITLE
does carbon support type erasure for Comparable?

### DISCUSSION THREAD

**Discussion by fililili** [carbon-language/carbon-lang#4535] [Getting started] [answered]
hi, all. I found Swift supports such usage for type erasure. It's interesting for me and I am looking for other language that supports same usage. So, Is such usage supported by carbon, too? thanks.
```Swift
func selfc<T: Comparable>(_ a: T) -> Bool {
    print(a)
    return a < a;
}

/* cannot compile
func selfca(_ a: any Comparable) -> Bool {
    return a < a;
}
*/

struct Person: Comparable {
    var name: String
    var age: Int
    
    static func < (lhs: Person, rhs: Person) -> Bool {
        return lhs.age < rhs.age
    }
}
let person1 = Person(name: "Alice", age: 30)
let person2 = Person(name: "Bob", age: 25)

var cs : [any Comparable] = [person1, person2, 1, 2.4]
for c in cs {
    print(selfc(c))
}
```

**Comment by danakj** [comment]
Yes, what you're describing is what Carbon calls checked generics. See https://github.com/carbon-language/carbon-lang/tree/trunk/docs/design/generics for the design.

**Reply by fililili** [reply]
hi, @danakj . In the Swift code, it creates a **list** of Comparable objects from **different** types `var cs : [any Comparable] = [person1, person2, 1, 2.4]`. But I have read the page [https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/overview.md](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/overview.md) and search for "type erasure", it doesn't show such usage.

**Reply by danakj** [reply, marked as answer]
Thanks, I think what you're looking for then is variadics with checked generics which is not finished but is mentioned here with some links: https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#variadic-arguments

Edit: If not, then maybe you're looking for a container, which would use checked generics as its type. That should follow from the original link, it may just not be called "type erasure" directly. But a checked generic can be used as a type for a container.

**Reply by geoffromer** [reply]
A checked generic type can be used as the element type of an array, but that still requires all elements of the array to have the same runtime type, just like with C++ templates. [Variadics](https://github.com/carbon-language/carbon-lang/pull/2240) will let you form a tuple whose elements have different runtime types that all implement some interface, and then iterate over them using that interface, but it's still a tuple, not an array. For example, you can't index into it with a runtime subscript.

If I understand the Swift example correctly, `var cs : [any Comparable] = [person1, person2, 1, 2.4]` is doing something different from either of those: it's forming an array, not a tuple, and its elements have different runtime types. The Carbon generics design refers to that kind of feature as "dynamic types", and it's currently marked as [future work](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#dynamic-types).

**Reply by fililili** [reply]
I found we can do same thing in c++ like below:
```cpp
class Comparable {
  class Value;
public:
  template <typename T>
  Comparable(const T& value, bool(*compare)(const T& t1, const T& t2))
    : _value {reinterpret_cast<const Value&>(value)},
      _compare {reinterpret_cast<bool(*)(const Value&, const Value&)>(compare)} {
  }
  template <typename Functor>
  auto callFunctor(Functor func) {
    return func.template useComparable<Value>(_value, _compare);
  }
private:
  const Value& _value;
  bool(*_compare)(const Value&, const Value&);
};

struct selfCompareFunctor {
  template <typename T>
  bool useComparable(const T& t, bool(*compare)(const T& t1, const T& t2)) {
    return compare(t, t);
  }
};
bool compareInt(const int& i1, const int& i2) {
  return i1 < i2;
}
int main() {
  int i{1};
  Comparable comp{i, compareInt};
  comp.callFunctor(selfCompareFunctor{});
}
```
In the code, I use incomplete type `Value` to represent any type. And user can write any functor they like and use callFunctor to call it.
What's more, it's enough with type check, functor can be instantized by incomplete type, means the template functor support all types.
And if user want associate type, they can use different incomplete type to distinguish different associate type.
Will carbon do same thing like this?

**Reply by geoffromer** [reply]
Yes, you can do type erasure in C++ (it's how `std::function` works, for example), so Carbon will need to support it in some form, but we haven't worked out the specifics. My guess is there will eventually be some higher-level memory-safe "dynamic types" feature where the language takes care of the mechanics, and a lower-level feature that can be an interop and migration target for C++ code like yours, but that doesn't have a language-level guarantee of memory safety.

**Reply by fililili** [reply]
Thanks for reply. But let me emphasize, the mechanism is powerful than std::function.
The key idea of the example is:
We should not use `type_erased_obj.virtual_function` but use  `type_erased_obj.call_functor(generic_function)` direclty.
`type_erased_obj.call_functor(generic_function)` is more type safe and natural, `type_erased_obj.virtual_function` is just a Syntactic Sugar of `type_erased_obj.call_functor(generic_function)`.
For example, in rust, it wrongly choose `type_erased_obj.virtual_function`. So if rust want to do selfCompare, it has to call dyn_trait_obj.compare(dyn_trait_obj) which is not type safe. But in my example, we just call rust `comp.callFunctor(selfCompareFunctor{});` which is always type safe. That's why rust has object safety (or call it dyn compatible) limitation.
And here is a more cleared version with comment. If the generic function has same binary no matter what type is, then the generic function can be used dynamic. Which should be easy to satisfied in Carbon.
```cpp
#include <bit>
#include <iostream>

template <typename Value>
struct Comparable {
  const Value& value;
  bool(*compare)(const Value&, const Value&);
};

class ErasedComparable {
  class Value;
public:
  template <typename T>
  ErasedComparable(Comparable<T> comparable)
    : _comparable {std::bit_cast<Comparable<Value>, Comparable<T>>(comparable)} {
  }
  template <typename Functor>
  auto callFunctor(Functor func) {
    return func.template useComparable<Value>(_comparable);
  }
private:
  Comparable<Value> _comparable;
};

struct selfCompareFunctor {
  template <typename T>
  bool useComparable(Comparable<T> comparable) {
    return comparable.compare(comparable.value, comparable.value);
  }
};

bool compareInt(const int& i1, const int& i2) {
  std::cout << "compare " << i1 << " and " << i2
            << " , the result is: " << (i1 < i2) << std::endl;
  return i1 < i2;
}

int main() {
  int i{1};
  Comparable<int> comparable{i, compareInt};

  // When the real type of 'comparable' is known, no cast is needed, and it is always type-safe.
  selfCompareFunctor{}.useComparable(comparable);

  // useComparable<double> and useComparable<int> are completely identical at the binary level.
  // Therefore, forcibly invoking useComparable<double> on a Comparable<int> object
  // can also produce the correct result.
  selfCompareFunctor{}.useComparable<double>(
    std::bit_cast<Comparable<double>>(comparable)
  );

  // However, exposing such casts publicly is too risky, so we encapsulate them using ErasedComparable.
  // Only the constructor of ErasedComparable and the callFunctor function are exposed to users.
  // As long as the code compiles:
  // 1. The constructor is constrained, so ErasedComparable can only be successfully constructed
  //    from some particular Comparable<T>.
  // 2. func.template useComparable<Value> can be instantiated successfully. 'Value' is a private,
  //    incomplete type, so func.useComparable<int>, func.useComparable<double>,
  //    and func.useComparable<Value> are completely identical at the binary level.
  // Therefore, within callFunctor, calling func.useComparable<Value> with a Comparable<Value>
  // that was cast from an unknown Comparable<T> will always produce the correct result.
  // In addition, _comparable is a private member and can only be accessed via the safe callFunctor.
  ErasedComparable erasedComparable{comparable};
  erasedComparable.callFunctor(selfCompareFunctor{});
}
```
